### PR TITLE
Adds new substring function to str plugin

### PR DESCRIPTION
**Pull request by Flare576**
Full disclosure: I've never written Rust before.

Function added but doesn't perform action yet, hoping for feedback. Code annotated with intentions since actual code might not reflect goal due to misunderstanding.

Do not merge this PR; code compiles and runs but does not add new functionality yet, causing test to fail.

### CODE REVIEW

File: src/plugins/str.rs [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
-    SyntaxShape, Tagged, TaggedItem, Value,
+    SyntaxShape, Tagged, Tag, TaggedItem, Value,
 };
 
 #[derive(Debug, Eq, PartialEq)]
 enum Action {
     Downcase,
     Upcase,
     ToInteger,
+    Substring,
```
Comment (jonathandturner) [inline src/plugins/str.rs]: You might want to go ahead and store the parameter for Substring in this enum case, so when you match later you can get it back out again

Comment (jonathandturner) [reply src/plugins/str.rs]: You can see some examples of how to do this in the WebEvent at the top of: https://doc.rust-lang.org/rust-by-example/custom_types/enum.html

File: src/plugins/str.rs [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
 }
 
 pub type ColumnPath = Vec<Tagged<String>>;
 
 struct Str {
     field: Option<ColumnPath>,
-    params: Option<Vec<String>>,
+    // I believe this should now support "named properties", like --substring "x,y"...
+    params: Vec<Tagged<String>>,
```
Comment (jonathandturner) [inline src/plugins/str.rs]: If you store the substring start/finish positions in the enum, I think you don't have to do this step

File: src/plugins/str.rs [hunk 3]
```diff
@@ -148,6 +168,25 @@ impl Plugin for Str {
...
+                    Tagged {
+                        item: Value::Primitive(Primitive::String(s)),
+                        tag,
+                    } => {
+                        self.params.push(s.clone().tagged(tag));
+                    }
+                    _ => {}
+                }
+            }
+            self.for_substring();
```
Comment (jonathandturner) [inline src/plugins/str.rs]: once you get the param figured out, then you could pass it as an argument to your `for_substring` call, and then build the enum value there

File: src/plugins/str.rs [hunk 4]
```diff
@@ -33,6 +35,15 @@ impl Str {
         let applied = match self.action.as_ref() {
             Some(Action::Downcase) => Value::string(input.to_ascii_lowercase()),
             Some(Action::Upcase) => Value::string(input.to_ascii_uppercase()),
+            /*
+              This line was a test to see if anything is being pushed to self.params... it isn't
+              Some(Action::Substring) => Value::string(self.params[0].item.to_string()),
+              The following is just to avoid crashing on execution
+            */
+            Some(Action::Substring) => Value::string(input),
```
Comment (jonathandturner) [inline src/plugins/str.rs]: Pattern matching lets you get the parameter for substring back out here, so you can use it to substr input

File: docs/commands/str.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ # │   │ name       │ path
+───┼───┼────────────┼────────────────────────────────
+ 0 │ X │ filesystem │ /home/tux/stuff/expr/stuff
+ 1 │   │ filesystem │ /
+━━━┷━━━┷━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+> shells | str # --to-int
+━━━┯━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+ # │   │ name       │ path
+───┼───┼────────────┼────────────────────────────────
+ 0 │ X │ filesystem │ /home/TUX/stuff/expr/stuff
```
Comment (jonathandturner) [inline docs/commands/str.md]: I wonder if we should use a different example for --to-int

File: src/plugins/str.rs [hunk 6]
```diff
@@ -33,6 +35,26 @@ impl Str {
         let applied = match self.action.as_ref() {
             Some(Action::Downcase) => Value::string(input.to_ascii_lowercase()),
             Some(Action::Upcase) => Value::string(input.to_ascii_uppercase()),
+            Some(Action::Substring(s)) => {
+                // Index operator isn't perfect: https://users.rust-lang.org/t/how-to-get-a-substring-of-a-string/1351
+                let no_spaces: String = s.chars().filter(|c| !c.is_whitespace()).collect();
```
Comment (jonathandturner) [inline src/plugins/str.rs]: You could also split, and then run .trim() on the results of the split

File: src/plugins/str.rs [hunk 7]
```diff
@@ -33,6 +35,26 @@ impl Str {
...
+                    _ => v[0].parse().unwrap(),
+                };
+                let end: usize = match v[1] {
+                    "" => input.len(),
+                    _ => cmp::min(v[1].parse().unwrap(), input.len()),
+                };
+                if start > input.len() - 1 {
+                    Value::string("")
+                } else if start > end {
+                    Value::string(input)
```
Comment (jonathandturner) [inline src/plugins/str.rs]: I wonder if we should return an error in the case where start comes after end.

File: docs/commands/str.md (hunk 5 above)
Comment (Flare576) [reply docs/commands/str.md]: I believe you're right, I'm not exactly sure what a good example is, though; when does `--to-int` shine? I had a hard time thinking of a case when converting to an int would make a significant difference to a following operation.

Comment (jonathandturner) [reply docs/commands/str.md]: You could use the `sum` command, since it will need integers to sum together.

Comment (jonathandturner) [reply docs/commands/str.md]: Like if you do something like:

```
echo "1, 2, 3" | split-row "," | str --to-int | sum
```

After that `split-row` you get strings back, and you can't sum them until you do the `str --to-int` over the column.

File: src/plugins/str.rs (hunk 7 above)
Comment (Flare576) [reply src/plugins/str.rs]: For the life of me I could not figure out how to test an error coming out of `Plugin::filter`. Every de-structuring/unwrapping/crying/pleading I tried failed... Then I realized that's not where it belongs anyway and now we have a much more logical approach to sub-parameters!

File: src/plugins/str.rs (hunk 6 above)
Comment (Flare576) [reply src/plugins/str.rs]: Way more clean, especially in the new split-focus approach

**Comment by jonathandturner** [comment]
Thanks!

**Pull request merged**